compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5390,6 +5395,28 @@ private BType checkInvocationArgs(BLangInvocation iExpr, List<BType> paramTypes,
         }
     }
 
+    private void checkArrayLibSortFuncArgs(BLangInvocation iExpr) {
+        if (iExpr.argExprs.size() <= 2) {
+            if (!types.isBasicType(iExpr.expr.type)) {
+                dlog.error(iExpr.expr.pos, DiagnosticCode.INVALID_SORT_ARRAY_MEMBER_TYPE, iExpr.expr.type);
+            }
+        } else if (iExpr.argExprs.size() == 3) {
+            if (iExpr.argExprs.get(2).type != null) {
+                if (iExpr.argExprs.get(2).type.tag == TypeTags.SEMANTIC_ERROR) {
+                    return;
+                } else if (iExpr.argExprs.get(2).type.tag == TypeTags.NIL && !types.isBasicType(iExpr.expr.type)) {
+                    dlog.error(iExpr.expr.pos, DiagnosticCode.INVALID_SORT_ARRAY_MEMBER_TYPE, iExpr.expr.type);
+                }
+            } else {
+                BLangLambdaFunction func = (BLangLambdaFunction) iExpr.argExprs.get(2);
+                BType returnType = func.function.type.getReturnType();
+                if (!types.isBasicType(returnType)) {
+                    dlog.error(iExpr.expr.pos, DiagnosticCode.INVALID_SORT_FUNC_RETURN_TYPE, returnType);
+                }
+            }
+        }
+    }
```

Comment:
There shouldn't be any need to hard code the type checking for this. Why do we need to do this?

Reply:
In the spec issue, James has mentioned that
"The compiler needs to check a few things that cannot be expressed in the signature:

- if a key function is specified as an argument, then its return value must be an ordered type

- if there is no key function, then the member type of the array must be ordered type"

https://github.com/ballerina-platform/ballerina-spec/issues/572#issuecomment-663838997